src/plugins/visualizations/public/utils/saved_visualize_utils.ts

```diff
@@ -0,0 +1,392 @@
+    throw new SavedObjectNotFound(SAVED_VIS_TYPE, id || '');
+  }
+
+  savedObject._source = _.cloneDeep(resp.attributes);
+
+  if (savedObject._source.visState) {
+    savedObject._source.visState = JSON.parse(savedObject._source.visState);
+  }
+
+  // assign the defaults to the response
+  _.defaults(savedObject._source, defaultsProps);
+
+  Object.assign(savedObject, savedObject._source);
+  savedObject.lastSavedTitle = savedObject.title;
+
+  savedObject.sharingSavedObjectProps = {
+    aliasTargetId,
+    outcome,
+    errorJSON:
+      outcome === 'conflict' && services.spaces
+        ? JSON.stringify({
+            targetType: SAVED_VIS_TYPE,
+            sourceId: id,
+            targetSpace: (await services.spaces.getActiveSpace()).id,
+          })
+        : undefined,
+  };
+
+  const meta = savedObject._source.kibanaSavedObjectMeta || {};
+
+  if (meta.searchSourceJSON) {
+    try {
+      let searchSourceValues = parseSearchSourceJSON(meta.searchSourceJSON);
+
+      if (opts.searchSource) {
+        searchSourceValues = injectSearchSourceReferences(
+          searchSourceValues as any,
+          resp.references
+        );
+        savedObject.searchSource = await services.search.searchSource.create(searchSourceValues);
+      } else {
+        savedObject.searchSourceFields = searchSourceValues;
+      }
+    } catch (error: any) {
+      throw error;
+    }
+  }
+
+  if (resp.references && resp.references.length > 0) {
+    config.injectReferences(savedObject, resp.references);
+  }
+
+  if (services.savedObjectsTagging) {
+    savedObject.tags = services.savedObjectsTagging.ui.getTagIdsFromReferences(resp.references);
+  }
+
+  savedObject.visState = await updateOldState(savedObject.visState);
+  if (savedObject.searchSourceFields?.index) {
+    await services.dataViews.get(savedObject.searchSourceFields.index as any);
+  }
+
+  return savedObject;
+}
+
+export async function saveVisualization(
+  savedObject: VisSavedObject,
+  {
+    confirmOverwrite = false,
+    isTitleDuplicateConfirmed = false,
+    onTitleDuplicate,
+  }: {
+    confirmOverwrite?: boolean;
+    isTitleDuplicateConfirmed?: boolean;
+    onTitleDuplicate?: () => void;
+  },
+  services: {
+    savedObjectsClient: SavedObjectsClientContract;
+    chrome: ChromeStart;
+    overlays: OverlayStart;
+    savedObjectsTagging?: SavedObjectsTaggingApi;
+  }
+) {
+  // Save the original id in case the save fails.
+  const originalId = savedObject.id;
+  // Read https://github.com/elastic/kibana/issues/9056 and
+  // https://github.com/elastic/kibana/issues/9012 for some background into why this copyOnSave variable
+  // exists.
+  // The goal is to move towards a better rename flow, but since our users have been conditioned
+  // to expect a 'save as' flow during a rename, we are keeping the logic the same until a better
+  // UI/UX can be worked out.
+  if (savedObject.copyOnSave) {
+    delete savedObject.id;
+  }
+
+  const attributes: any = {
```

Comment:
Sorry, but `any` should be removed. We already have all into to correctly typescriptify that. 
This type can be reused not only here but from  all methods above 🔆

Reply:
typescriptified that